src/Keboola/ExEmailAttachments/Action/RunAction.php

```diff
@@ -171,10 +172,9 @@ public function saveFiles(array $userConfiguration, array $files) : void
         }
 
         // Get header from first file
-        $firstFileHandle = fopen($files[0], 'r');
-        $firstFileHeader = fgetcsv($firstFileHandle);
-        fclose($firstFileHandle);
-
+        $firstFile = new CsvFile($files[0], $userConfiguration['delimiter'], $userConfiguration['enclosure']);
```

Comment:
tady bych se bál, jestli nezůstane otevřený ten soubor! Normálně s těma souborama nějak pracuješ, klidně v cyklu, ale jakmile se odreferencuje ta proměnná, tak se to v desktruktoru zavře. Ale tady ta proměnná je skrz celý proces a dole se nad tím pouští tail - který možná nebude mít přístup. Testoval jsi to pls?

Reply:
Aha, jasně. Testama to projde: https://github.com/keboola/ex-email-attachments/blob/master/tests/Keboola/ExEmailAttachments/Unit/RunActionTest.php#L194 Je teda pravda, že do kódu `CsvFile` jsem nekoukal, ale když jsem nenašel v našeptávači `close()`, tak už jsem to neřešil.

Reply:
To jsem myslím rozporoval v review toho CSV - že by bylo fajn mít extra "close" metodu.

Reply:
Ok, takže jak to mám teda udělat, vrátit se k `fgetcsv()`? ;o)

Reply:
Mělo by stačit unset($firstFile)

Reply:
Tím se zavolá destruktor (s největší pravděpodobností, byť se na to nedá spoléhat) a mělo by se to zavřít. Nebo můžeme počkat, jestli se to vysype. Možná to bude v pohodě.

Reply:
No to právě vůbec není jistý, ne?

Reply:
Ale tak ok, minimálně se tím nic nezkazí.

Reply:
https://github.com/keboola/php-csv/issues/33

Reply:
Ono celé by se to dost možná mělo řešit úplně jinak přes procesory a ne tady dole přes tail. Co myslíš @ondrejhlavacek ?

Reply:
V S3 extraktoru se to řeší tak, že se hlavička načte procesorem `processor-create-manifest` z random souboru a pak se přes `processor-skip-lines` udělá tail. Ale ty procesory tam transparentně hodí UI, takže dost záleží, jaký UI je plánovaný tady.

Reply:
"chyba" když to tak řeknu... mi tady přijde, že ta komponenta je hrozně moc navázaná na to jak funguje KBC. Má za zodpovědnost víc věcí, než by měla/musela. Její jedinou zodpovědností by mělo být konverze: [S3 bucket s emaily] -> [csv soubory v přílohách emailů]

A to je vše. Všechno ostatní by měly řešit jiné komponenty/procesory. (jen IMHO)